### DISCUSSION TITLE
关于存储方式的咨询

### DISCUSSION THREAD

**Discussion by SuoSiFire** [milvus-io/milvus#16581] [Q&A and General discussion] [answered]
我目前使用的是milvus v1.1.1版本，由于公司要求，需要把写文件的方式改成网络传输。
想请教下是不是 模仿 https://github.com/milvus-io/milvus/blob/1.x/core/src/storage/s3/S3IOWriter.cpp 这里封装一个类似的类就好。
通过这个类看上去是把要写的内容缓存到内存，然后一起上传。 
想了解下milvus是不是没有追加写的场景，所以只要缓存进内存全量写文件就可以？
最后再确认下，除了 https://github.com/milvus-io/milvus/blob/1.x/core/src/storage这里还有其他设计文件读写的代码吗？

**Comment by yhmo** [comment, marked as answer]
是的，可以仿照s3下面的代码来写。主要是要实现IOWriter/IOReader/Operation这几个类里声明的接口。
core/src/storage/FSHandler.h 里面的createFsHandler()函数在运行时根据设置产生不同的IOWriter/IOReader/Operation实例。
milvus中所有的IO操作(包括读写文件，新建/删除文件/文件夹)都依赖这套接口。理论上你只用实现你的IOWriter/IOReader/Operation子类，然后在createFsHandler()加一个分支即可。

追加写的情况是存在的，比如每个segment文件夹里有一个delete_doc文件，用来记录被删除的entities，当客户端调用delete()删除某个entity时，milvus就会去改写这个delete_doc文件。对于这种情况，参考core/src/codecs/default/DefaultDeletedDocsFormat.cpp里面的write()，是这么处理的：
1. 调用Operation::CacheGet(file_path)，如果是本地存储，这个接口什么都不做，如果是s3，这个接口会把文件下载到本地
2. 正常调用open() write() close()对本地文件改写
3. 调用Operation::CachePut(file_path)，如果是本地存储，这个接口什么都不做，如果是s3，这个接口会把文件上传覆盖

**Reply by SuoSiFire** [reply]
好的，感谢回复，我尝试改动下。

**Comment by SuoSiFire** [comment]
麻烦再咨询一个问题，milvus磁盘文件中单个文件大小最大有多大？ 是和index_file_size相关吗？  @yhmo

**Reply by SuoSiFire** [reply]
顺便再咨询下，接口内部是否需要封装重试逻辑，比如某次网络问题导致的put/get失败，需要在接口内重试直到成功吗？还是业务侧会做对应的手段？ @yhmo

**Reply by yhmo** [reply]
milvus分片文件的大小主要跟index_file_size有关，这不是一个精确值，理论上最大的分片小于index_file_size的1.999999倍，一般情况下是比index_file_size略大。
put/get失败在接口实现里重试吧，如果业务侧读不到数据，就直接报错了。

**Reply by SuoSiFire** [reply]
好的，接口里面重试应该有一定次数限制的，如果重试多次后依旧失败，是否会导致数据异常呢？ 读节点读不到数据的情况可以接受，主要是担心写节点写数据失败导致数据异常。

**Reply by yhmo** [reply]
在1.x里，就比如磁盘满了，写不下去，就会在日志里报错。如果wal写成功了，清理出磁盘空间后重启服务，数据会重新写。但如果wal没写成功，insert就会返回错误，相当于数据没insert成功。
你重新实现接口也是类似，如果你的接口里网络传输重试都失败了，也只能日志里报错。那数据还在wal里，重启之后如果网络还是没好，那一样还是没法恢复数据。

**Reply by SuoSiFire** [reply]
好的，了解了，那是没问题的。感谢回复。

**Comment by SuoSiFire** [comment]
按照上述方法改写后，对于集合执行milvus.io中提供的example.py是可以正常运行并且结果正确的。 但是当再次对集合进行插入操作时，就会core dump。gdb了一下但感觉不太可能在这一行core dump，麻烦帮忙看下日志有没有哪里能看出问题。 万分感谢。 @yhmo 

info日志（中间加了些debug用的日志）：
[milvus-22-05-17-23_33-info.log](https://github.com/milvus-io/milvus/files/8709672/milvus-22-05-17-23_33-info.log)

debug日志，最后一行日志打印后程序就core dump了
[milvus-22-05-17-23_33-debug.log](https://github.com/milvus-io/milvus/files/8709650/milvus-22-05-17-23_33-debug.log)

gdb日志，定位在CacheConfigHandler.cpp:83这一行 core dump
[05170test.log](https://github.com/milvus-io/milvus/files/8709663/05170test.log)

![image](https://user-images.githubusercontent.com/90120870/168853243-8d6f30b3-907a-4b2d-b9ab-e3428d32f226.png)

这里我打印了相关变量都是正常的，麻烦帮忙看下可能是什么原因? @yhmo

**Comment by SuoSiFire** [comment]
@yhmo  这个有空帮忙看下吗，谢谢

**Reply by yhmo** [reply]
看起来这个位置不应该会crash。Config::GetInstance()是个单例。没看出来什么原因

**Reply by SuoSiFire** [reply]
> 看起来这个位置不应该会crash。Config::GetInstance()是个单例。没看出来什么原因

是的，这里比较奇怪，core dump的位置也不太一致，感觉上是堆栈内存异常导致的core dump。但是用valgrind测了下没发现什么异常。

**Reply by yhmo** [reply]
你做了哪些代码改动？

**Reply by SuoSiFire** [reply]
这个问题貌似已经定位到了，我在实现读写类时有一处字符串拷贝写的有问题，导致内存异常了。 感谢回复，后续有问题再来问。

**Comment by SuoSiFire** [comment]
@yhmo  麻烦再确认一个问题，使用远端存储时，wal文件是存储在写节点本地还是存在远端的。

**Reply by yhmo** [reply]
wal写在本地，代码里是直接读写本地磁盘。wal需要不停地写文件删文件，如果存在远端，会使得insert操作很迟钝。

**Reply by SuoSiFire** [reply]
了解，如果禁用wal，对于服务来说影响是什么？是写入性能的影响还是可能会丢数据？  @yhmo

**Reply by SuoSiFire** [reply]
目前的问题是，服务起在k8s集群内，并且集群不能挂载宿主机的磁盘，导致每次重启后wal文件就丢了。 担心这样会有数据丢失，想问下s3场景下是怎么解决这个问题的？ 或者有什么建议吗？ @yhmo

**Reply by yhmo** [reply]
1.x的wal只能在本地磁盘读写，没法设置成远程读写，所以必须要有本地磁盘才能工作，这是个避不开的限制。
